### PR TITLE
Update logical rules when pipeline parallelism is used

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

Previously, we directly fix the logical rules from config when calling sharding hints in certain modules, e.g. attention_op and moe. We notice it brings troubles since we use a different a sharding rule for certain layers when using pipeline parallelism. This PR fixes this issue.

FIXES: b/484417665

# Tests

We leverage the debug_sharding feature and applied a train_compile simulation using tpu7x-1024 on deepseek model, see output: https://paste.googleplex.com/6497058040184832. 

The debug sharding information suggests the pipeline-specifc rule is successfully applied, e.g.
```
I0302 18:01:56.977987 127929547775552 attention_mla.py:805] attention_mla/q_pe Physical: bfloat16[1024,2048,128,64].................................. (('stage', 'fsdp'), None, 'tensor', None).
I0302 18:01:56.995703 127929547775552 attention_mla.py:809] attention_mla/query Logical: bfloat16[1024,2048,128,192]................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:56.995775 127929547775552 attention_mla.py:809] attention_mla/query Physical: bfloat16[1024,2048,128,192]................................. (('stage', 'fsdp'), None, 'tensor', None).
I0302 18:01:57.022124 127929547775552 attention_mla.py:830] attention_mla/key_nope Logical: bfloat16[1024,2048,128,128]................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:57.022189 127929547775552 attention_mla.py:830] attention_mla/key_nope Physical: bfloat16[1024,2048,128,128]................................. (('stage', 'fsdp'), None, 'tensor', None).
I0302 18:01:57.039021 127929547775552 attention_mla.py:831] attention_mla/key_rope Logical: bfloat16[1024,2048,128,64].................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:57.039082 127929547775552 attention_mla.py:831] attention_mla/key_rope Physical: bfloat16[1024,2048,128,64].................................. (('stage', 'fsdp'), None, 'tensor', None).
I0302 18:01:57.056057 127929547775552 attention_mla.py:835] attention_mla/key Logical: bfloat16[1024,2048,128,192]................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:57.056114 127929547775552 attention_mla.py:835] attention_mla/key Physical: bfloat16[1024,2048,128,192]................................. (('stage', 'fsdp'), None, 'tensor', None).
I0302 18:01:57.073061 127929547775552 attention_mla.py:836] attention_mla/value Logical: bfloat16[1024,2048,128,128]................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:57.073130 127929547775552 attention_mla.py:836] attention_mla/value Physical: bfloat16[1024,2048,128,128]................................. (('stage', 'fsdp'), None, 'tensor', None).
I0302 18:01:57.128059 127929547775552 attention_op.py:1640] attention_op/query Logical: bfloat16[1024,128,2048,192]................................. Unknown.
I0302 18:01:57.128147 127929547775552 attention_op.py:1640] attention_op/query Physical: bfloat16[1024,128,2048,192]................................. (('stage', 'fsdp'), 'tensor', None, None).
I0302 18:01:57.145525 127929547775552 attention_op.py:1641] attention_op/key Logical: bfloat16[1024,128,2048,192]................................. Unknown.
I0302 18:01:57.145588 127929547775552 attention_op.py:1641] attention_op/key Physical: bfloat16[1024,128,2048,192]................................. (('stage', 'fsdp'), 'tensor', None, None).
I0302 18:01:57.162914 127929547775552 attention_op.py:1642] attention_op/value Logical: bfloat16[1024,128,2048,128]................................. Unknown.
I0302 18:01:57.162975 127929547775552 attention_op.py:1642] attention_op/value Physical: bfloat16[1024,128,2048,128]................................. (('stage', 'fsdp'), 'tensor', None, None).
I0302 18:01:57.210172 127929547775552 attention_mla.py:1058] attention_mla/out Logical: bfloat16[1024,2048,128,128]................................. ('activation_batch', 'activation_length_no_exp', 'activation_heads', 'activation_kv').
I0302 18:01:57.210253 127929547775552 attention_mla.py:1058] attention_mla/out Physical: bfloat16[1024,2048,128,128]................................. (('stage', 'fsdp'), None, 'tensor', None).
I0302 18:01:57.227630 127929547775552 deepseek.py:225] deepseek/attention_result Logical: bfloat16[1024,2048,7168].................................... ('activation_batch', 'activation_norm_length', 'activation_embed').
I0302 18:01:57.227692 127929547775552 deepseek.py:225] deepseek/attention_result Physical: bfloat16[1024,2048,7168].................................... (('stage', 'fsdp'), None, 'tensor').
I0302 18:01:57.246183 127929547775552 deepseek.py:200] deepseek/post_attention_norm Logical: bfloat16[1024,2048,7168].................................... ('activation_batch', 'activation_norm_length', 'activation_embed').
I0302 18:01:57.246245 127929547775552 deepseek.py:200] deepseek/post_attention_norm Physical: bfloat16[1024,2048,7168].................................... (('stage', 'fsdp'), None, 'tensor').
```
It means DP is applied for dense/out-of-pipeline layers! For pipeline layers, activations are sharded properly
```
I0302 18:01:59.264940 127929547775552 attention_mla.py:830] attention_mla/key_nope Physical: bfloat16[128,2048,128,128].................................. ('fsdp', None, 'tensor', None).
I0302 18:01:59.282920 127929547775552 attention_mla.py:831] attention_mla/key_rope Logical: bfloat16[128,2048,128,64]................................... ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:59.282977 127929547775552 attention_mla.py:831] attention_mla/key_rope Physical: bfloat16[128,2048,128,64]................................... ('fsdp', None, 'tensor', None).
I0302 18:01:59.301332 127929547775552 attention_mla.py:835] attention_mla/key Logical: bfloat16[128,2048,128,192].................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:59.301392 127929547775552 attention_mla.py:835] attention_mla/key Physical: bfloat16[128,2048,128,192].................................. ('fsdp', None, 'tensor', None).
I0302 18:01:59.319334 127929547775552 attention_mla.py:836] attention_mla/value Logical: bfloat16[128,2048,128,128].................................. ('activation_kv_batch', 'activation_length_no_exp', 'activation_kv_heads', 'activation_kv_head_dim').
I0302 18:01:59.319396 127929547775552 attention_mla.py:836] attention_mla/value Physical: bfloat16[128,2048,128,128].................................. ('fsdp', None, 'tensor', None).
I0302 18:01:59.346043 127929547775552 attention_op.py:1640] attention_op/query Logical: bfloat16[128,128,2048,192].................................. Unknown.
I0302 18:01:59.346114 127929547775552 attention_op.py:1640] attention_op/query Physical: bfloat16[128,128,2048,192].................................. ('fsdp', 'tensor', None, None).
I0302 18:01:59.364462 127929547775552 attention_op.py:1641] attention_op/key Logical: bfloat16[128,128,2048,192].................................. Unknown.
I0302 18:01:59.364524 127929547775552 attention_op.py:1641] attention_op/key Physical: bfloat16[128,128,2048,192].................................. ('fsdp', 'tensor', None, None).
I0302 18:01:59.382726 127929547775552 attention_op.py:1642] attention_op/value Logical: bfloat16[128,128,2048,128].................................. Unknown.
I0302 18:01:59.382801 127929547775552 attention_op.py:1642] attention_op/value Physical: bfloat16[128,128,2048,128].................................. ('fsdp', 'tensor', None, None).
I0302 18:01:59.435019 127929547775552 attention_mla.py:1058] attention_mla/out Logical: bfloat16[128,2048,128,128].................................. ('activation_batch', 'activation_length_no_exp', 'activation_heads', 'activation_kv').
I0302 18:01:59.435103 127929547775552 attention_mla.py:1058] attention_mla/out Physical: bfloat16[128,2048,128,128].................................. ('fsdp', None, 'tensor', None).
I0302 18:01:59.454271 127929547775552 deepseek.py:225] deepseek/attention_result Logical: bfloat16[128,2048,7168]..................................... ('activation_batch', 'activation_norm_length', 'activation_embed').
I0302 18:01:59.454331 127929547775552 deepseek.py:225] deepseek/attention_result Physical: bfloat16[128,2048,7168]..................................... ('fsdp', None, 'tensor').
I0302 18:01:59.475671 127929547775552 deepseek.py:200] deepseek/post_attention_norm Logical: bfloat16[128,2048,7168]..................................... ('activation_batch', 'activation_norm_length', 'activation_embed').
I0302 18:01:59.475736 127929547775552 deepseek.py:200] deepseek/post_attention_norm Physical: bfloat16[128,2048,7168]..................................... ('fsdp', None, 'tensor').
I0302 18:01:59.498431 127929547775552 moe.py:1410] moe/inputs Logical: bfloat16[128,2048,7168]..................................... ('activation_batch', 'activation_norm_length', None).
I0302 18:01:59.498498 127929547775552 moe.py:1410] moe/inputs Physical: bfloat16[128,2048,7168]..................................... ('fsdp', None, None).
I0302 18:01:59.516621 127929547775552 moe.py:1411] moe/gate_logits Logical: bfloat16[128,2048,256]...................................... ('activation_batch', 'activation_norm_length', None).
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/attention_op.py [hunk 1]
```diff
@@ -595,21 +579,15 @@ def maybe_create_nnx(einsum, *args):
       self.AqtEinsum_3 = jnp.einsum
 
   def _logical_to_mesh_axes(self, logical_name):
-    return logical_to_mesh_axes(
-        logical_name, mesh=self.mesh, rules=self.config.logical_axis_rules
-    )
+    logical_rules = None if self.config.using_pipeline_parallelism else self.config.logical_axis_rules
```
Comment (gobbleturk) [inline src/maxtext/layers/attention_op.py:582]: Ideally the logical axis rules should be plumbed explicitly through and never be none, but this LGTM as its an improvement over the current code =)

Comment (NuojCheng) [reply src/maxtext/layers/attention_op.py:582]: I might prefer setting different logical names, e.g. `activation_batch, ['dp', 'fsdp']` and `activation_batch_out_pipeline, ['stage', 'dp', 'fsdp']`, but it is also not easy to update. Right now let's rely on context manager to insert correct rules.
